R/parallel.R

```diff
@@ -65,42 +67,109 @@ test_files_parallel <- function(
     load_package = load_package
   )
 
-  files <- list()
-
-  with_reporter(reporters$multi,
-    while (!queue$is_idle()) {
-      # TODO: poll for 100-200ms to be able to update a spinner
-      msgs <- queue$poll(Inf)
-      for (x in msgs) {
-        if (x$code != PROCESS_MSG) {
-          next
-        }
-
-        m <- x$message
-        if (!inherits(m, "testthat_message")) {
-          message(m)
-          next
-        }
-
-        # Record all events until we get end of file, then we replay them all
-        # with the local reporters. This prevents out of order reporting.
-        if (m$cmd != "DONE") {
-          files[[m$filename]] <- append(files[[m$filename]], list(m))
-        } else {
-          replay_events(reporters$multi, files[[m$filename]])
-          reporters$multi$end_context_if_started()
-          files[[m$filename]] <- NULL
-        }
-      }
+  with_reporter(reporters$multi, {
+    parallel_update <- reporter$capabilities$parallel_update
+    if (parallel_update) {
+      parallel_event_loop_smooth(queue, reporters)
+    } else {
+      parallel_event_loop_chunky(queue, reporters)
     }
-  )
+  })
 
   test_files_check(reporters$list$get_results(),
     stop_on_failure = stop_on_failure,
     stop_on_warning = stop_on_warning
   )
 }
 
+default_num_cpus <- function() {
+  # Use common option, if set
+  ncpus <- getOption("Ncpus", NULL)
+  if (!is.null(ncpus) && !is_integer(ncpus)) {
+    stop("`getOption(Ncpus)` must be integer")
+  }
+
+  if (!is.null(ncpus)) return(ncpus)
+
+  # Otherwise detect
+  ncpus <- ps::ps_cpu_count()
```

Comment:
```suggestion
  ncpus <- ps::ps_cpu_count() - 1L
```
Just to be polite?

Reply:
Actually, I think it is completely fine to take all CPUs. If this behavior bothers you because you regularly run some CPU-intensive programs on your dev machine, then you can set the new env var to cap the CPUs?

Reply:
Ok, fine with me.